### PR TITLE
fix(server): require at least one field to be set when updating memory

### LINKED ISSUE #27833: Malformed PUT /api/memories api call leads to internal server error
### I have searched the existing issues, both open and closed, to make sure this is not a duplicate report.

- [x] Yes

### The bug

Attempting to update a memory via `/api/memories/{memory-id}` with a body not containing any of the specified values e.g. `isSaved`, `memoryAt` or `seenAt` causes a malformed sql update statement `sql: 'update "memory" set  where "id" = $1'` and returns an 

```JSON
{"message":"Failed to update memory","error":"Internal Server Error","statusCode":500,"correlationId":"6i774gka"}
```

I was trying to see if i could manipulate the other fields on my immich server and stumbled on this error.

This should be replicable via the demo or a vanilla immich instance with an existing memory.

### The OS that Immich Server is running on

Linux 16ac84d276d1 6.12.15-production+truenas #1 SMP PREEMPT_DYNAMIC Mon May 26 13:44:31 UTC 2025 x86_64 GNU/Linu

### Version of Immich Server

v2.5.6

### Version of Immich Mobile App

N/a

### Platform with the issue

- [x] Server
- [x] Web
- [ ] Mobile

### Device make and model

_No response_

### Your docker-compose.yml content

```YAML
Not relevant.
```

### Your .env content

```Shell
Not relevant.
```

### Reproduction steps

1. Get any existing memory via `/api/memories`
2. Create a put request on that memory. (Can be done in browser console.)
```JAVASCRIPT
await fetch("/api/memories/e6f619f1-9270-43e7-aec6-40dd7d9d3a06", {
  method: "PUT",
  body: JSON.stringify({}),
  headers: {
    "Content-type": "application/json; charset=UTF-8"
  }
});
```
Get a nice 500 Internal server error and an error message in the logs.

### Relevant log output

```shell
Query failed : {
2026-04-15 20:24:14.934944+00:00durationMs: 1.438970997929573,
2026-04-15 20:24:14.934968+00:00error: PostgresError: syntax error at or near "where"
2026-04-15 20:24:14.934989+00:00at ErrorResponse (/usr/src/app/server/node_modules/.pnpm/postgres@3.4.8/node_modules/postgres/cjs/src/connection.js:815:30)
2026-04-15 20:24:14.935009+00:00at handle (/usr/src/app/server/node_modules/.pnpm/postgres@3.4.8/node_modules/postgres/cjs/src/connection.js:489:6)
2026-04-15 20:24:14.935031+00:00at Socket.data (/usr/src/app/server/node_modules/.pnpm/postgres@3.4.8/node_modules/postgres/cjs/src/connection.js:324:9)
2026-04-15 20:24:14.935056+00:00at Socket.emit (node:events:508:28)
2026-04-15 20:24:14.935076+00:00at addChunk (node:internal/streams/readable:559:12)
2026-04-15 20:24:14.935097+00:00at readableAddChunkPushByteMode (node:internal/streams/readable:510:3)
2026-04-15 20:24:14.935119+00:00at Readable.push (node:internal/streams/readable:390:5)
2026-04-15 20:24:14.935141+00:00at TCP.onStreamRead (node:internal/stream_base_commons:189:23) {
2026-04-15 20:24:14.935163+00:00severity_local: 'ERROR',
2026-04-15 20:24:14.935184+00:00severity: 'ERROR',
2026-04-15 20:24:14.935206+00:00code: '42601',
2026-04-15 20:24:14.935230+00:00position: '22',
2026-04-15 20:24:14.935250+00:00file: 'scan.l',
2026-04-15 20:24:14.935271+00:00line: '1188',
2026-04-15 20:24:14.935293+00:00routine: 'scanner_yyerror'
2026-04-15 20:24:14.935314+00:00},
2026-04-15 20:24:14.935336+00:00sql: 'update "memory" set  where "id" = $1',
2026-04-15 20:24:14.935358+00:00params: [ '8be6bd49-1f66-415b-a70c-8e15c5c4f9b4' ]
2026-04-15 20:24:14.935391+00:00}
2026-04-15 20:24:14.935556+00:00[31m[Nest] 40  - [39m04/15/2026, 10:24:14 PM [31m  ERROR[39m [33m[Api:ErrorInterceptor~99t84fqt][39m [31mUnknown error: PostgresError: syntax error at or near "where"[39m
2026-04-15 20:24:14.935645+00:00PostgresError: syntax error at or near "where"
2026-04-15 20:24:14.935680+00:00at ErrorResponse (/usr/src/app/server/node_modules/.pnpm/postgres@3.4.8/node_modules/postgres/cjs/src/connection.js:815:30)
2026-04-15 20:24:14.935707+00:00at handle (/usr/src/app/server/node_modules/.pnpm/postgres@3.4.8/node_modules/postgres/cjs/src/connection.js:489:6)
2026-04-15 20:24:14.935734+00:00at Socket.data (/usr/src/app/server/node_modules/.pnpm/postgres@3.4.8/node_modules/postgres/cjs/src/connection.js:324:9)
2026-04-15 20:24:14.935760+00:00at Socket.emit (node:events:508:28)
2026-04-15 20:24:14.935786+00:00at addChunk (node:internal/streams/readable:559:12)
2026-04-15 20:24:14.935812+00:00at readableAddChunkPushByteMode (node:internal/streams/readable:510:3)
2026-04-15 20:24:14.935837+00:00at Readable.push (node:internal/streams/readable:390:5)
2026-04-15 20:24:14.935862+00:00at TCP.onStreamRead (node:internal/stream_base_commons:189:23)
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by fredfloydd**
## Description

Adds a nonEmptyPartial zod utility in src/validation.ts - works like z.object().partial() but rejects objects where every field is undefined. 

Uses it on MemoryUpdateSchema so empty PUT bodies are caught at the validation layer before reaching the service.

Fixes #27833

## How Has This Been Tested?

Before this change, PUT with an empty body to `/api/memories/{id}` returns a 500.

After this change, the same request returns a 400.

Requests with content still return 200.

## Checklist:

- [x] I have carefully read CONTRIBUTING.md
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [ ] I have confirmed that any new dependencies are strictly necessary.
- [ ] I have written tests for new code (if applicable)
- [ ] I have followed naming conventions/patterns in the surrounding code
- [ ] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [ ] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

## Please describe to which degree, if any, an LLM was used in creating this pull request.

I used an LLM to help write the curls to reproduce the issue and test the fix. The code execution is my own.

### CODE REVIEW

**Review by jrasm91 — CHANGES_REQUESTED** [review]
This doesn't work everywhere because plenty of repository methods are used outside of API requests so you can't just throw http exceptions. Almost all API validation happens in zod interceptor level before getting to the service so I would prefer to add it there if possible.

I'd be fine with updating the types although that may or may not match runtime so idk how helpful it is in reality.

**Comment by fredfloydd** [comment]
@jrasm91 yeah fair enough - there is precedent for other helpers in utils/ returning a bad request, but this is a bit different...

Pretty sure zod doesn't support the idea of 'set at least one key in this object' out of the box, but can probably define a refine() util for it somewhere.

I do still think the repository methods should be able to guard against queries that will fail, but that's a bit of a separate scope/less obviously causing issues

File: server/src/validation.ts [hunk 1]
```diff
@@ -32,6 +32,22 @@ export function IsIPRange(options?: IsIPRangeOptions) {
...
+ * @example
+ * nonEmptyPartial({ name: z.string(), bio: z.string() }).meta({ id: 'UpdateDto' });
+ */
+export function nonEmptyPartial<T extends z.ZodRawShape>(shape: T) {
+  return z
+    .object(shape)
+    .partial()
+    .refine((data) => Object.values(data as Record<string, unknown>).some((value) => value !== undefined), {
+      message: 'At least one field must be provided',
+    });
```
Comment (danieldietzler) [inline server/src/validation.ts:48]: Can't you check if `Object.keys(data).length !== 0`?

Comment (fredfloydd) [reply server/src/validation.ts:48]: Would that not let the input `{ x: undefined, y: undefined }` through?

Comment (danieldietzler) [reply server/src/validation.ts:48]: It would. Does that still cause postgres to error?

Comment (jrasm91) [reply server/src/validation.ts:48]: If there aren't any field being updated then an update would generate invalid SQL, yes

Comment (fredfloydd) [reply server/src/validation.ts:48]: Exactly - need zod to prevent this case passing through.

**Review by jrasm91 — COMMENTED** [review]
Yup, this is what I had in mind and it looks great. Now that I am looking at it I am just wondering whether it is better to throw a validation error like or have it do a noop and return early with a 200 success instead. That seems like it would be better for external callers, who don't have to worry about the case where all the updated fields might be undefined.

**Comment by fredfloydd** [comment]
Potentially, though I think there are a couple of decent arguments against.

If we return a 200, it could be a bit misleading if we haven't actually executed a sql query in the database - the updated_at wouldn't be changed for example.

Also, as in your first point, it's a bit of a code smell to do the validation in the memory service when it's already basically entirely handled by zod in the middleware.

If you think genuinely better to 200 then happy to go with that, but if I were a consumer I would _probably_ rather get a bad request response here (it also highlights that they're doing something a bit weird...)

**Comment by jrasm91** [comment]
Yeah, those are all good points. Let me ask the rest of the team and we'll go with whatever the majority is. I think I vote to throw an error.

**Comment by jrasm91** [comment]
Looks like the consensus is to throw an error

**Review by jrasm91 — APPROVED** [review]
Do you mind adding a memory controller test for the empty update?

File: server/src/validation.ts (hunk 1 above)
Comment (mertalev) [reply server/src/validation.ts:48]: In fairness `{ x: undefined, y: undefined }` is impossible from the API since you can't get an object like that after JSON serialization/deserialization.

Comment (fredfloydd) [reply server/src/validation.ts:48]: Yeah that's true, though I think current impl is still fine given that ultimately the intent of the schema should be to only pass things that will produce valid sql, rather than be most optimal for this case where we rely on implicit some implicit properties of the input.

Happy for me to resolve this comment?

**Comment by fredfloydd** [comment]
Done!

**Review by jrasm91 — APPROVED** [review]
Sweet ty
